Mathlib/Computability/ContextFreeGrammar.lean

```diff
@@ -6,14 +6,14 @@ Authors: Martin Dvorak
 import Mathlib.Computability.Language
 
 /-!
-# Context-Free Grammar
+# Context-Free Grammars
 
 This file contains the definition of a context-free grammar, which is a grammar that has a single
 nonterminal symbol on the left-hand side of each rule.
 
 ## Main definitions
 * `ContextFreeGrammar`: A context-free grammar.
-* `ContextFreeGrammar.language`: A language generated by a given context-free grammar.
```

Comment:
I think you want to keep both this and the new definition in the docstring

Reply:
OK, I'll put it back.
And the definitions through which `ContextFreeGrammar.language` is defined shouldn't be here?

Reply:
I am confused about the philosophy of "Main definitions". When we have a pile of definitions, do we mention only the top one? Because that was almost what was here before. I don't mind including them all tho.

Reply:
It's supposed to do what it says on the tin. Sometimes you have a lot of `def`s (eg you just defined a new space and are defining a bunch of maps from it) but only a few are important. Those are the ones that should be in the module doc.

Reply:
So here I should mention `ContextFreeGrammar.Produces`, `ContextFreeGrammar.Derives`, `ContextFreeGrammar.Generates`, `ContextFreeGrammar.language` and `Language.IsContextFree` as well?

Reply:
I think you should have
```
* `ContextFreeGrammar`: A context-free grammar.
* `ContextFreeGrammar.language g`: The language generated by the given context-free grammar `g`
* `Language.IsContextFree l`: The proposition that the language `l` is generated by some `ContextFreeGrammar`
```
Part of the problem with the current version is that that you are using the description of `language` to describe `IsContextFree`.

Reply:
Elaborating on what @YaelDillies said, why not `ContextFreeGrammar.Produces` and `ContextFreeGrammar.Derives` and `ContextFreeGrammar.Generates` as main definitions, too?

Reply:
Possibly I misunderstood "what it says on the tin".

Reply:
I think of `Produces`/`Derives`/`Generates` all as secondary definitions to `ContextFreeGrammar`.

Reply:
Isn't `ContextFreeGrammar.language` a secondary definition to `Language.IsContextFree` then?

Reply:
No, because without that definition you would never have a reason to talk about `ContectFreeGrammar` at all.

Reply:
I don't understand the philosophy here. In my point of view `Language.IsContextFree` is built on top of `ContextFreeGrammar.language` as well as `ContextFreeGrammar.language` is built on top of `ContextFreeGrammar.Generates` and so on.

Reply:
It's not about what things are built upon; if I tell you that I have constructed a `ContextFreeGrammar`, what would you guess is the most likely thing I want to do with it? My guess is that the answer is "look at the language it defines".

Reply:
Yes!

Reply:
I better do this PR without `Language.IsContextFree` as it should be about prefix and postfix.